src/blockrelay/graphene.cpp

```diff
@@ -82,6 +82,126 @@ void CGrapheneBlock::FillShortTxIDSelector()
     shorttxidk1 = shorttxidhash.GetUint64(1);
 }
 
+void CGrapheneBlock::AddNewTransactions(std::vector<CTransaction> vMissingTx, CNode *pfrom)
+{
+    if (vMissingTx.size() == 0)
+        return;
+
+    // If canonical ordering is activated, locate empty indexes in vTxHashes256 to be used in sorting
+    std::vector<size_t> missingTxIdxs;
+    if (fCanonicalTxsOrder && NegotiateGrapheneVersion(pfrom) >= 1)
+    {
+        uint256 nullhash;
+        for (size_t idx = 0; idx < vTxHashes256.size(); idx++)
+        {
+            if (vTxHashes256[idx] == nullhash)
+                missingTxIdxs.push_back(idx);
+        }
+    }
+
+    size_t idx = 0;
+    for (const CTransaction &tx : vMissingTx)
+    {
+        mapMissingTx[GetShortID(pfrom->gr_shorttxidk0, pfrom->gr_shorttxidk1, tx.GetHash(),
+            NegotiateGrapheneVersion(pfrom))] = MakeTransactionRef(tx);
+
+        uint256 hash = tx.GetHash();
+        uint64_t cheapHash =
+            GetShortID(pfrom->gr_shorttxidk0, pfrom->gr_shorttxidk1, hash, NegotiateGrapheneVersion(pfrom));
+
+        // Insert in arbitrary order if canonical ordering is enabled and xversion is recent enough
+        if (fCanonicalTxsOrder && NegotiateGrapheneVersion(pfrom) >= 1)
+            vTxHashes256[missingTxIdxs[idx++]] = hash;
```

Comment:
please put idx++ on its own line so the order is clear.  This post-incrementor is a c-ism that's notorious for creating bugs.

Also, I'm concerned about idx exceeding missingTxIdxs.size().  I'd like to see this checked within this routine to keep the check near the use.  

Did you mean to prevent this case on line 241?  That check doesn't do it because missingTxIdxs is the size of the null items in vTxHashes256 not its full size.

Reply:
I moved idx++ to a separate line and also added two range checks. The first makes sure vMissingTx.size() == missingTxIdxs.size() outside the loop. The second, which should be redundant, makes sure that idx does not exceed the capacity of missingTxIdxs.size().